Gamma ray propagation

andrewfullard opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

**Description**
<!--- Describe your changes in detail -->

This cleans up the previous PR and consists of:
* Montecarlo code to propagate decay products in ejecta (photons only)
* Handling of immediate positron decay
* Some unit tests
* A little documentation

**Motivation and context**
<!--- Why is this change required? What problem does it solve? Link issues here -->

This code is the first step in late-phase SNe Ia simulations in TARDIS

**How has this been tested?**
- [x] Testing pipeline.
- [ ] Other. <!--- please describe how you tested your changes, `pytest` flags used, etc. -->

**Examples**
<!-- If appropiate, link notebooks, screenshots and other demo stuff -->
Notebook included in PR
**Type of change**
<!--- Put an `x` in all the boxes that apply -->
- [ ] Bug fix. <!-- non-breaking change which fixes an issue -->
- [x] New feature. <!-- non-breaking change which adds functionality -->
- [ ] Breaking change. <!-- fix or feature that would cause existing functionality to not work as expected -->
- [ ] None of the above. <!-- please describe -->

**Checklist**
<!--- Put an `x` in all the boxes that apply -->
- [x] My change requires a change to the documentation.
    - [ ] I have updated the documentation accordingly.
    - [ ] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [x] I have assigned and requested two reviewers for this pull request.

andrewfullard committed:
Radioactive decay product propagation

This cleans up the previous PR and consists of:
* Montecarlo code to propagate decay products in ejecta (photons only)
* Handling of immediate positron decay
* Some unit tests
* A little documentation

andrewfullard committed:
Radioactive decay product propagation

This cleans up the previous PR and consists of:
* Montecarlo code to propagate decay products in ejecta (photons only)
* Handling of immediate positron decay
* Some unit tests
* A little documentation

andrewfullard committed:
Docstrings

andrewfullard committed:
Docstrings

afloers committed:
reworked gamma-ray deposition

afloers committed:
reworked gamma-ray deposition

andrewfullard committed:
Fixes and improvements for pair creation

andrewfullard committed:
Fixes and improvements for pair creation

andrewfullard committed:
Hidden a lot of prints, more explicit positron opacity energies

andrewfullard committed:
Hidden a lot of prints, more explicit positron opacity energies

andrewfullard committed:
Fixed some function calls, incorporated X-rays

andrewfullard committed:
Fixed some function calls, incorporated X-rays

andrewfullard committed:
More vector normalization

andrewfullard committed:
More vector normalization

andrewfullard committed:
Fix E-R formula, little cleanup

andrewfullard committed:
Fix E-R formula, little cleanup

andrewfullard committed:
Existing tests fixed

andrewfullard committed:
Existing tests fixed

andrewfullard committed:
Removed prints

andrewfullard committed:
Removed prints

andrewfullard committed:
Renamed GammaRay object to GXPacket, refactoring

andrewfullard committed:
Renamed GammaRay object to GXPacket, refactoring

andrewfullard committed:
More renaming and refactoring

andrewfullard committed:
More renaming and refactoring

andrewfullard committed:
Removed unneeded tau subtraction

andrewfullard committed:
Removed unneeded tau subtraction

andrewfullard committed:
Improved deposition time calculation

Now returns values in seconds

andrewfullard committed:
Improved deposition time calculation

Now returns values in seconds

andrewfullard committed:
Added energy deposition time of zero for beta decay

Also output energy deposition time

andrewfullard committed:
Added energy deposition time of zero for beta decay

Also output energy deposition time

afloers committed:
added decay scheme using tardis nuclear and streamlined spawning process

afloers committed:
added decay scheme using tardis nuclear and streamlined spawning process

andrewfullard committed:
Added todo comment

andrewfullard committed:
Added todo comment

andrewfullard committed:
Refactoring

Changed energy output to a dataframe.

andrewfullard committed:
Refactoring

Changed energy output to a dataframe.

andrewfullard committed:
Calculate energy per shell per gram

andrewfullard committed:
Calculate energy per shell per gram

andrewfullard committed:
Docstrings and cleanup

andrewfullard committed:
Docstrings and cleanup

andrewfullard committed:
Cleaned up unused imports

andrewfullard committed:
Cleaned up unused imports

andrewfullard committed:
Fixed main loop docstring

andrewfullard committed:
Fixed main loop docstring

andrewfullard committed:
Added calculated iron group fraction and log packet index in plotting dataframe

andrewfullard committed:
Added calculated iron group fraction and log packet index in plotting dataframe

andrewfullard committed:
Fix moving packets to interaction being slightly too far

andrewfullard committed:
Fix moving packets to interaction being slightly too far

andrewfullard committed:
Addressing most comments

andrewfullard committed:
Addressing most comments

andrewfullard committed:
Fixec circular dependency and other minor items

andrewfullard committed:
Fixec circular dependency and other minor items

andrewfullard committed:
Docstring for escaping energy output

andrewfullard committed:
Docstring for escaping energy output

andrewfullard committed:
Merge branch 'gamma_ray_propagation' of github.com:andrewfullard/tardis into gamma_ray_propagation

andrewfullard committed:
Added chvogl comments as TODO comments in code.

andrewfullard committed:
Increased boundary crossing factor to prevent boundary sticking

andrewfullard committed:
Refactoring of CDFs and improved compton angle sampling

Simplified a number of functions so fewer are operating on packet objects.

andrewfullard committed:
Restructuring and cleanup. Some better tests

andrewfullard committed:
Tests now pass with refdata

andrewfullard commented:
Tests will pass once https://github.com/tardis-sn/tardis-refdata/pull/47 is merged

andrewfullard committed:
Normalized energy deposition rate to decay rate

wkerzendorf commented on docs/physics/gamma_ray_deposition.rst:
```diff
@@ -0,0 +1,9 @@
+Gamma Ray Energy Deposition
+***************************
+
+This notebook provides the initial implementation of Gamma-ray energy deposition into an arbitrary ejecta.
```
I think this can just be in the notebook

wkerzendorf commented on docs/physics/gammaray_deposition.ipynb:
```diff
@@ -0,0 +1,573 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #4.    from tardis.energy_input.base import GXPacket, main_gamma_ray_loop
</pre><p>at some stage we should rename <code>GXPacket</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #12.    #config.model.structure.velocity.stop = 2e4 * u.km / u.s
</pre><p>remove comments?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #3.    
</pre><p>maybe some description about the plot would be great (goes for all of the plots)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #4.    velocity = (np.array(ejecta_energy_r)/(config.supernova.time_explosion * u.day.to(u.s))) * u.cm.to(u.km)
</pre><p>multiply <code>ejecta_r</code>  with <code>u.cm</code>  and then just do the conversion at the end (<code>.to(u.km/u.s)</code>)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #3.    ax.hist(np.array(ejecta_energy_r)/np.max(model.v_outer[:].value), bins=200)
</pre><p>labels on axis and description on plot. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    np.sum((model.density * model.volume)).to(u.M_sun)
</pre><p>maybe just put a comment that this is the ejecta mass.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #3.    plt.plot(0,0)
</pre><p>labels and comments</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    sum(escape_energy) / (sum(escape_energy) + sum(ejecta_energy))
</pre><p>we should use <code>np.sum</code>  and some description</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

wkerzendorf replied:
<pre class="ql-syntax" spellcheck="false">Line #4.    ax.semilogy();
</pre><p>axis labels and description, remove comments</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

wkerzendorf replied:
<p>what are the columns here - a description</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

wkerzendorf replied:
<p>that is very clear plot and nice!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1698/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+    ----------
+    number_of_shells : int64
+        Number of shells in model
+    packets_per_shell : pandas dataframe
```
for IDEs to pick-up on this stuff actually give them a valid python expression - `pandas.DataFrame`

wkerzendorf commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,143 @@
+    float
+        Fractional intensity of source_2
+    """
+    intensity_1 = get_type_property(nuclear_data, source_1, "intensity")
```
the referees felt that the name `get_type_property` had a name that is a bit ambiguous (but failed to come up with a better name).

wkerzendorf commented:
some of the docstrings don't quite follow the numpydoc format.

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+        )
+        for shell in range(number_of_shells):
+            required_packets_in_shell = packets_per_shell[column].iloc[shell]
+            for i_packet in range(required_packets_in_shell):
```
`i_packet` is a bit confusing name hree.

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+            required_packets_in_shell = packets_per_shell[column].iloc[shell]
+            for i_packet in range(required_packets_in_shell):
+                # draw a random gamma-ray in shell
+                ray1 = GXPacket(0, 0, 1, GXPacketStatus.IN_PROCESS, 0)
```
`ray1` maybe should be `primary_photon`

chvogl replied:
Having named arguments would be useful here. What `0, 0, 1, GXPacketStatus.IN_PROCESS, 0` means is not immediately clear.

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                # draw a random gamma-ray in shell
+                ray1 = GXPacket(0, 0, 1, GXPacketStatus.IN_PROCESS, 0)
+                z1 = np.random.random()
+                initial_radius = inner_velocities[shell] + z1 * (
```
we should do `1/r^3`.

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                ray1.location = SphericalVector(
+                    initial_radius, location_theta, location_phi
+                )
+                direction_theta = get_random_theta_gamma_ray()
```
we should make sure that we make this conform with TARDIS.

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                    1.0, direction_theta, direction_phi
+                )
+                ray1.shell = shell
+                z2 = np.random.random()
```
not sure if we need a new variable

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                )
+                ray1.shell = shell
+                z2 = np.random.random()
+                if z2 > gamma_ratio:
```
```suggestion
                if gamma_ratio < np.random.random():
```

chvogl replied:
It would be useful to have a comment here explaining which physical scenario this corresponds to (positron annihilation ...).

afloers replied:
gamma_ratio is not the best name for this. How about 'gamma_ray_probability'?

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                        ray1_y,
+                        ray1_z,
+                    ) = ray1.direction.cartesian_coords
+                    ray2 = copy.deepcopy(ray1)
```
```suggestion
                    secondary_photon = copy.deepcopy(ray1)
```

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+    # Note the use of velocity as the radial coordinate
+    inner_radius = model.v_inner[0].value
+    outer_radius = model.v_outer[-1].value
+    outer_velocities = model.v_outer[:].value
```
is that a copy? if so `model.v_outer.value.copy()` is cleaner

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+    )
+
+    # Taking iron group to be elements 21-30
+    iron_group_fraction_per_shell = raw_isotope_abundance.loc[(21,):(30,)].sum(
```
we should make sure that we document that this is a choice because of atomic data.

chvogl commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                z2 = np.random.random()
+                if z2 > gamma_ratio:
+                    # positron: sets gamma-ray energy to 511keV
+                    ray1.energy = 511.0
```
I would make this a named constant.

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+
+    i = 0
+    for packet in tqdm(packets):
+        interaction_count.append(0)
```
remove

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+
+                if packet.status == GXPacketStatus.PAIR_CREATION:
+                    ejecta_energy_gained = packet.energy - (2.0 * 511.0)
+                    packet, backward_ray = pair_creation(packet)
```
rename photon

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+
+                # Save packets to dataframe rows
+                energy_df_rows[packet.shell] += (
+                    ejecta_energy_gained * 1000.0 / ejecta_volume[packet.shell]
```
document why its times 1000

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+            else:
+                packet.tau -= total_opacity * distance_boundary * ejecta_epoch
+                # overshoot so that the gamma-ray is comfortably in the next shell
+                packet = move_gamma_ray(packet, distance_boundary * (1 + 1e-7))
```
we should use a constant here - settable in the config

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+                )
+                packet.shell = get_shell(packet.location.r, outer_velocities)
+
+            if (
+                np.abs(packet.location.r - outer_radius) < 10.0
+                or packet.shell > len(ejecta_density) - 1
+            ):
+                escape_energy.append(packet.energy)
+                packet.status = GXPacketStatus.END
+            elif (
+                np.abs(packet.location.r - inner_radius) < 10.0
+                or packet.shell < 0
+            ):
+                packet.energy = 0.0
+                packet.status = GXPacketStatus.END
```
remove and see if it works

wkerzendorf commented on tardis/energy_input/base.py:
```diff
@@ -0,0 +1,369 @@
+    energy_df_rows *= scaled_decay_rate_per_shell
+
+    energy_df = pd.DataFrame(
+        data=energy_df_rows, columns=["energy [eV/s/cm^-3]"]
```
maybe just explicitely document multiple times in the code

wkerzendorf commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,127 @@
+from tardis import constants as const
+from tardis.energy_input.util import kappa_calculation
+
+MASS_SI = 28.085 * const.m_p.to(u.g).value
```
use existing atom data.

wkerzendorf replied:
`AtomData.atom_data`

wkerzendorf commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,127 @@
+        The energy of the packet
+    ejecta_density : float
+        The density of the ejecta
+
```
add latex string here.

wkerzendorf replied:
look for reference

wkerzendorf commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,127 @@
+    float
+        Photoabsorption opacity
+    """
+    Si_opacity = (
```
lower case

wkerzendorf commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,127 @@
+    float
+        Pair creation opacity
+    """
+    Z_Si = 14
```
from atom_data

wkerzendorf commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,127 @@
+        + Fe_proton_ratio * iron_group_fraction
+    )
+
+    if energy > 1.022e3 and energy < 1.5e3:
```
document source of numbers

wkerzendorf commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,143 @@
+# Assume photon path is small compared to dynamical effects
+
+
+def read_nuclear_dataframe(path):
```
should be in `nuclear` or removed

wkerzendorf commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,143 @@
+    One-dimensional Numpy Array, dtype float
+        Array of values matching the queried type and property
+    """
+    return nuclear_df.query("type==" + type_of_radiation)[property].values
```
does this need to be a function - not sure.

wkerzendorf commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,143 @@
+import numpy as np
+import pandas as pd
+
+# Need to scale deposited energy to rate of decay reaction to get energy per second in steady state
```
look over comments

chvogl replied:
```suggestion

```

wkerzendorf commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,143 @@
+    float
+        Sampled energy
+    """
+    z = np.random.random()
```
maybe remove this variable

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    return distance_interaction, distance_boundary
+
+
+def move_gamma_ray(gxpacket, distance):
```
move photons?

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    )
+
+
+def solve_quadratic_equation(x, y, z, x_dir, y_dir, z_dir, radius_velocity):
```
maybe more similar to this https://github.com/tardis-sn/tardis/blob/master/tardis/montecarlo/montecarlo_numba/calculate_distances.py#L21 - also in name

chvogl commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    inner_radii : One dimensional Numpy array, dtype float
+    outer_radii : One dimensional Numpy array, dtype float
+    total_opacity : dtype float
+    ejecta_epoch : dtype float
```
```suggestion
    ejecta_epoch : float
```

chvogl commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    gxpacket : GXPacket object
+    inner_radii : One dimensional Numpy array, dtype float
+    outer_radii : One dimensional Numpy array, dtype float
+    total_opacity : dtype float
```
```suggestion
    total_opacity : float
```

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+import tardis.constants as const
+
+
+def calculate_distance_radial(gxpacket, r_inner, r_outer):
```
use https://github.com/tardis-sn/tardis/blob/master/tardis/montecarlo/montecarlo_numba/calculate_distances.py#L21 as a template to restructure this and make more clear for which cases

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+        outer_radii[gxpacket.shell],
+    )
+
+    distance_interaction = gxpacket.tau / total_opacity / ejecta_epoch
```
`ejecta_epoch=time_explosion`?

chvogl commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+
+    Parameters
+    ----------
+    radii : GammaRay object
```
Is `radii` really a `GammaRay object`?

wkerzendorf replied:
the type might be wrong 😄

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    return gxpacket
+
+
+def density_sampler(radii, mass_ratio):
```
is mass_ratio abundances

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+
+    Returns
+    -------
+    radius : dtype float
```
descriptions

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+
+    mass_ratio_sorted_indices = np.argsort(mass_ratio)
+    index = mass_ratio_sorted_indices[
+        np.searchsorted(mass_ratio, z, sorter=mass_ratio_sorted_indices)
```
is `mass_ratio`s a cumulative distribution function

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    int
+        Shell index corresponding to radius
+    """
+    shell_inner = np.searchsorted(outer_radii, radius, side="left")
```
is this function needed - or can we inline

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    norm_shell_masses = shell_masses / np.sum(shell_masses)
+    abundance_dict = {}
+    nuclide_mass_dict = {}
+    for index, row in raw_isotope_abundance.iterrows():
```
```suggestion
    for (atom_number, atom_mass), row in raw_isotope_abundance.iterrows():
```

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+    decay_rate_per_shell_df = nuclide_mass_df.copy()
+    for column in activity_df:
+        isotope_meta = meta.loc[column]
+        half_life = isotope_meta.loc[
+            isotope_meta["key"] == "Parent T1/2 value"
+        ]["value"].values[0]
```
this seems reasonably complicated. will this change with nuclear?

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+        half_life = isotope_meta.loc[
+            isotope_meta["key"] == "Parent T1/2 value"
+        ]["value"].values[0]
+        half_life = convert_half_life_to_astropy_units(half_life)
```
this should already be in seconds?

wkerzendorf commented on tardis/energy_input/gamma_ray_grid.py:
```diff
@@ -0,0 +1,226 @@
+            isotope_meta["key"] == "Parent T1/2 value"
+        ]["value"].values[0]
+        half_life = convert_half_life_to_astropy_units(half_life)
+        atomic_mass = float(re.findall("\d+", column)[0])
```
we should use the atomic mass from the Atom database

chvogl commented on tardis/energy_input/gamma_ray_interactions.py:
```diff
@@ -0,0 +1,160 @@
+
+    Returns
+    -------
+        GXPacket
+            forward packet
+        GXPacket
+            backward packet
```
```suggestion
GXPacket
    forward packet
GXPacket
     backward packet
```

wkerzendorf commented on tardis/energy_input/gamma_ray_interactions.py:
```diff
@@ -0,0 +1,160 @@
+        final_compton_scattered_vector[2],
+    )
+
+    return theta_final.value + 0.5 * np.pi, phi_final.value
```
why do we always add 90 degrees

andrewfullard replied:
Because of the astropy spherical/cartesian conversion axis choice

wkerzendorf commented on tardis/energy_input/gamma_ray_interactions.py:
```diff
@@ -0,0 +1,160 @@
+    direction_theta = get_random_theta_gamma_ray()
+    direction_phi = get_random_phi_gamma_ray()
+
+    gxpacket.energy = 511.0
```
```suggestion
    gxpacket.energy = 511.0 #keV
```

wkerzendorf replied:
should be a named constant calculated from `astropy.constants`

wkerzendorf commented on tardis/energy_input/gamma_ray_interactions.py:
```diff
@@ -0,0 +1,160 @@
+    gxpacket.direction.theta = direction_theta
+    gxpacket.direction.phi = direction_phi
+
+    backward_ray = GXPacket(
```
why not do `deepcopy` on all of it?

wkerzendorf commented on tardis/energy_input/tests/test_energy_input_base.py:
```diff
@@ -0,0 +1,30 @@
+
+@pytest.mark.skip(reason="To be implemented")
+def test_initialize_packets():
+    assert False
```
do we still want to skip that?

wkerzendorf commented on tardis/energy_input/tests/test_gamma_ray_grid.py:
```diff
@@ -0,0 +1,66 @@
+    ----------
+    basic_gamma_ray : GammaRay object
+    """
+    gamma_ray = basic_gamma_ray
```
we should test easy cases (or edge cases) - move along a direction - move all directions - rather than all

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    END = 4
+
+
+class GXPacket(object):
```
this should be in its own module - called whatever we want to call GXPacket

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    kappa : float
+
+    """
+    k = energy / 511.0
```
constant

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    return k
+
+
+def euler_rodrigues(theta, direction):
```
use this https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html?

andrewfullard replied:
I don't know that it is identical so I would rather use the explicit calculation

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    Returns
+    -------
+    float
+        Klein-Nishina solution
```
maybe a link to wikipedia or something where we can see the formula - definitely add in latex to the doc string

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    float64
+        Energy density in erg / g
+    """
+    return energy * KEV2ERG / mass
```
inline function>

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    normalized_vector : One-dimensional Numpy Array, dtype float
+    """
+    normalized_vector = vector / np.sqrt(
+        vector[0] ** 2 + vector[1] ** 2 + vector[2] ** 2
```
this should be the array squared but probably better `normalized_v = v/np.linalg.norm(v)`

wkerzendorf commented on tardis/energy_input/util.py:
```diff
@@ -0,0 +1,333 @@
+    """
+    [value, unit] = half_life_string.split(" ")
+    try:
+        [nominal_value, magnitude] = value.split("×")
```
braces needed?

wkerzendorf reviewed (commented):
There is quite a lot of string parsing going on for the nuclear data which we might want to change.

the numpydoc is not consistent yet

otherwise very impressive addition to TARDIS

andrewfullard committed:
Addressing review comments

- Packet -> photon across the branch code
- Improved some docstrings
- Other cleanup as per comments

andrewfullard committed:
[build docs]

Addressed even more comments

andrewfullard committed:
[build docs]

andrewfullard committed:
Addressed more comments

andrewfullard committed:
Refactoring a little, addressed some comments

Moved GXphoton class to its own file

andrewfullard committed:
Corrected sampling of photon initial radius, and added nuclear to env

andrewfullard committed:
Added a simple pynonthermal connector that will solve the SF equation for each grid cell

andrewfullard committed:
Added scaled_decay_rate_per_shell output

andrewfullard committed:
Major update

Added astropy units across the majority of the module. Abundances are now correctly decayed to appropriate levels. Ensured correct sorting of energies and intensities for the energy CDF. Fixed calculations of nuclear activity.

andrewfullard committed:
Latest changes to scaling factors

andrewfullard committed:
Energy deposition comparison

andrewfullard committed:
Adding secondary gamma-ray from beta decay annihilation

andrewfullard committed:
Normalized decays by number of nuclides per shell

Also renamed some variables to make more sense

andrewfullard committed:
Fix dataframe copy fail, return to abundance-scaled decays

andrewfullard committed:
Removed quantity objects from internal calculations

andrewfullard committed:
Removed astropy coord transform

andrewfullard committed:
Decay product energy scaled by activity when created

andrewfullard committed:
Fixed Compton angle sampling, electron radius squared and photoabsorption opacity

andrewfullard committed:
Applied numba where easiest

andrewfullard committed:
Added frame transformations that are hopefully correct, improved some docstrings

andrewfullard committed:
Commentary on doppler effects, improved some tests

andrewfullard committed:
Cleaned up some output

andrewfullard committed:
Updated GXPhoton to be a jitclass, cleaned up Compton scattering

andrewfullard committed:
Frame transformation adjustments, renamed methods

andrewfullard committed:
Opacities to rest frame

andrewfullard committed:
Indivisible packet formalism version

andrewfullard committed:
Corrected frame transform in photon version

andrewfullard committed:
Pair creation frame transform freq after scatter

andrewfullard committed:
Correct use of packet frequency instead of energy to calculate opacities

andrewfullard committed:
Change to compton fraction calculation instead of energy loss

andrewfullard committed:
Added rf/cmf packet properties and using artis data

andrewfullard committed:
Possibly some progress

andrewfullard committed:
Corrected some doppler factors, correctly scaled activity per shell

andrewfullard committed:
Changes to opacities and transport to closely match artis

andrewfullard committed:
Returned to standard K-N formulation and added some useful comments

andrewfullard committed:
Small performance gain, using nuclear for isotopes

andrewfullard committed:
Added grey opacity option

andrewfullard committed:
Estimator scaffolding

andrewfullard committed:
Parallel gamma-ray loop, performance improvements

andrewfullard committed:
Added packet tracking back with good performance

andrewfullard committed:
Logging positrons again

andrewfullard committed:
Time dependence

andrewfullard committed:
More time dependent work

Rodot- commented on tardis/energy_input/indivisible_packets.py:
```diff
@@ -136,6 +137,7 @@ def initialize_packets(
             energy = co56_lines.energy.to_numpy() * 1000
             intensity = co56_lines.intensity.to_numpy()
+            # positron energy scaled by intensity
             positron_energy = 0.63 * 1000 * 0.19
```
Could you make these numbers named constants or something like that?

Rodot- commented on tardis/energy_input/indivisible_packets.py:
```diff
@@ -313,10 +315,11 @@ def main_gamma_ray_loop(num_decays, model):
-    ).sum() + activity_df["Co56"] * (
-        co56_lines.energy * 1000 * co56_lines.intensity
-    ).sum()
+        activity_df["Ni56"]
```
Make sure you have a comment describing the unit conversion each time you multiply by one of the 1000s (I know it should be obvious, but best to document all code magic numbers)

Rodot- commented on tardis/energy_input/indivisible_packets.py:
```diff
@@ -313,10 +315,11 @@ def main_gamma_ray_loop(num_decays, model):
+    )
 
     print("Total expected energy")
     print(total_energy.sum())
```
Reminder to remove these or convert to log messages at some point

andrewfullard committed:
Merge branch 'master' into gamma_ray_propagation
Get the new env into this branch

andrewfullard committed:
Merge branch 'master' into gamma_ray_propagation

andrewfullard committed:
Packets constant with time

andrewfullard committed:
Matching Urilight to 20 days or so

andrewfullard committed:
artis-type pair creation opacity and inverse volume with time

andrewfullard committed:
Added isotope_mass as a plasma property similar to atomic_mass

andrewfullard committed:
Added isotope_mass as a plasma property similar to atomic_mass

andrewfullard committed:
Added isotope number density calculation

andrewfullard committed:
Added isotope number density calculation

andrewfullard committed:
Added simple (currently failing) test

andrewfullard committed:
Added simple (currently failing) test

andrewfullard committed:
Working tests

andrewfullard committed:
Working tests

andrewfullard committed:
Merge branch 'plasma_isotope_properties' into gamma_ray_propagation

andrewfullard committed:
Revert "Merge branch 'plasma_isotope_properties' into gamma_ray_propagation"

This reverts commit d69f929b940c4bac21e9cc66d958925b099cc9a4, reversing
changes made to 93fa0c74ef3d135411386732cfc13a87ff7ad3a2.

andrewfullard committed:
Fixes for Compton fraction

andrewfullard committed:
Docstrings and cleanup

andrewfullard committed:
Corrected doppler factor vectors

andrewfullard committed:
Gamma packets to cartesian vector space

andrewfullard committed:
Fixed (probably) shell location after moving packet

andrewfullard committed:
Removed multi-threading

andrewfullard committed:
Fixed relativistic aberration returning non-unit vectors

andrewfullard committed:
Added expanding shell distance option, some cleanup

Rodot- commented on tardis/energy_input/indivisible_packets.py:
```diff
@@ -615,7 +615,7 @@ def process_packet_path(packet):
 
 
-@njit(parallel=True)
+@njit
```
Please set this to use the `njit_dict_no_parallel` from montecarlo/montecarlo_numba.  This does raise a good point that our numba config shouldn't be in the numba folder, but instead global.

andrewfullard committed:
Spectrum output, linear time step option, cleanup

andrewfullard committed:
Docstring cleanup

andrewfullard committed:
Cleaned out some old code, most tests pass again

andrewfullard committed:
Documentation

andrewfullard committed:
njit_dict_no_parallel applied, removed old unused functions

andrewfullard committed:
Applied black to new files

andrewfullard committed:
Merge branch 'master' into gamma_ray_propagation

andrewfullard committed:
Test fix to remove excess file changes

andrewfullard committed:
Fix excess PR changes

andrewfullard committed:
Final cleanup

andrewfullard committed:
Updated docs to support updates to astropy units

andrewfullard committed:
Split out nuclear dependence

andrewfullard committed:
Black, again

Rodot- reviewed (commented):
Looks good to me, I'll approve once black passes

andrewfullard committed:
Numba based packet init

andrewfullard committed:
Fix missing intensity multiplication

andrewfullard committed:
Updated docs, added WIP deposition estimator

andrewfullard committed:
Estimator sort of working

Rodot- commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,249 @@
+    )
+
+    fe_opacity = (
+        25.7e-24
```
Ditto

Rodot- commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,249 @@
+    float
+        Pair creation opacity
+    """
+    z_si = 14
```
Make this a global constant or get from atom data

Rodot- commented on tardis/energy_input/calculate_opacity.py:
```diff
@@ -0,0 +1,249 @@
+    float
+        Pair creation opacity
+    """
+    M_H = 1.67352e-24
```
Should be global

Rodot- commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,351 @@
+        CDF where each index corresponds to the energy in
+        the sorted array
+    """
+    energy.sort()
+    sorted_indices = np.argsort(energy)
```
sorted_indices will just be arange(len(energy)) right?

Rodot- commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,351 @@
+    intensity_1 = nuclear_data.query("type==" + source_1)["intensity"].values
+    intensity_2 = nuclear_data.query("type==" + source_2)["intensity"].values
+    total_intensity = np.sum(intensity_1) + np.sum(intensity_2)
+    scale_factor = total_intensity / 100
```
Add a comment here explaining the division by 100

Rodot- commented on tardis/energy_input/energy_source.py:
```diff
@@ -0,0 +1,351 @@
+        Energies and intensities of the isotope lines
+    """
+    return pd.read_csv(
+        "~/Downloads/tardisnuclear/" + isotope + ".txt",
```
Should this function take a path?

andrewfullard committed:
Address comments

andrewfullard committed:
Fix tests

andrewfullard committed:
More test fixes and docstrings

andrewfullard commented:
Some of the docstring failures are confusing. It claims 7 functions in e.g. calculate_opacity.py but there is definitely only 6.
